engine/src/board.rs

```diff
@@ -248,55 +275,71 @@ impl Board {
-        excluded_position: Position,
-        visited: &mut HashSet<Position>,
-    ) {
-        if visited.contains(&position) {
-            return;
-        }
-        visited.insert(position);
-        for pos in self.positions_taken_around_iter(position) {
-            if pos != excluded_position && !visited.contains(&pos) {
-                self.walk_board_inner(pos, excluded_position, visited);
-            }
+    pub fn update_pinned(&mut self) {
+        for pinned_info in self.calculate_pinned().iter() {
+            self.pinned[self.piece_to_offset(pinned_info.piece)] = pinned_info.pinned
         }
     }
 
-    pub fn pinned(&self, position: Position) -> bool {
-        // pieces on top of the hive cannot be pinned (just gated)
-        if self.level(position) > 1 {
-            return false;
-        }
-        // if there's only one neighbor the piece isn't pinned
-        let all_neighbor_positions = self
-            .positions_taken_around_iter(position)
+    pub fn calculate_pinned(&self) -> Vec<PinnedInfo> {
+        // make sure to get only top pieces in this
+        let mut ap_info = self
+            .positions
+            .iter()
+            .enumerate()
+            .filter_map(|(i, maybe_pos)| {
+                if let Some(pos) = maybe_pos {
+                    if self.is_top_piece(self.offset_to_piece(i), *pos) {
+                        Some(PinnedInfo {
+                            position: *pos,
+                            piece: self.top_piece(*pos).unwrap(),
+                            visited: false,
+                            depth: 0,
+                            low: 0,
+                            pinned: false,
+                            parent: None,
+                        })
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
             .collect::<Vec<_>>();
-        if all_neighbor_positions.len() < 2 {
-            return false;
+        if ap_info.len() == 0 {
+            return ap_info;
         }
-        let mut visited = HashSet::new();
-        visited = self.walk_board(*all_neighbor_positions.last().unwrap(), position, visited);
-        for neighbor_pos in all_neighbor_positions {
-            // if we can't reach all neighbors starting from a random neighbor, the piece would
-            // break the hive into two
-            if !visited.contains(&neighbor_pos) {
-                return true;
+        self.bcc(0, 0, &mut ap_info);
+        ap_info
+    }
+
+    pub fn bcc(&self, i: usize, d: usize, ap_info: &mut Vec<PinnedInfo>) {
+        ap_info[i].visited = true;
+        ap_info[i].depth = d;
+        ap_info[i].low = d;
+        let mut child_count = 0;
+        let mut ap = false;
+
+        for pos in self.positions_taken_around_iter(ap_info[i].position) {
+            let ni = ap_info.iter().position(|e| e.position == pos).unwrap();
```

Comment:
the unwrap on this line causes tests to panic
'board::tests::tests_positions_taken_around' panicked at 'called `Option::unwrap()` on a `None` value', engine/src/board.rs:325